### ISSUE TITLE
OpenEmu v1.0 isn't codesigned

### ISSUE THREAD

**Issue by lilyball** [OpenEmu/OpenEmu#816]
The OpenEmu v1.0 download is not codesigned, which means Gatekeeper will prevent the app from running by default.

**Comment by nekoniaow** [comment]
No need to speculate when the information is readily available:

From Apple's _Code Signing Overview_ (https://developer.apple.com/library/ios/documentation/Security/Conceptual/CodeSigningGuide/AboutCS/AboutCS.html#//apple_ref/doc/uid/TP40005929-CH3-SW3):

> All sorts of code can be signed, including tools, applications, scripts, libraries, plug-ins, and other “code-like” data.

and:

> To enable signed code to fulfill these purposes, a code signature consists of three parts:

A seal, which is a collection of checksums or hashes of the various parts of the code, such as the identifier, the Info.plist, the main executable, the resource files, and so on. The seal can be used to detect alterations to the code and to the app identifier.
A digital signature, which signs the seal to guarantee its integrity. The signature includes information that can be used to determine who signed the code and whether the signature is valid.
A unique identifier, which can be used to identify the code or to determine to which groups or categories the code belongs. This identifier can be derived from the contents of the Info.plist for the app, or can be provided explicitly by the signer.

So the info.plist does not have to be the part of the signature, it can, but that is not necessary.
As long as the code which checks that the info.plist is valid is itself properly part of the checksum-ed data then all is good.

**Comment by nekoniaow** [comment]
Is the fee the reason why this is not implemented?

**Comment by cyco** [comment]
Aside from the uncertainty about rewriting the info.plist, the fee is the main reason OpenEmu is not signed.

**Comment by nekoniaow** [comment]
Given how Mac users are usually way more prone than Windows/Linux ones to pay for stuff would it not be a good idea to pop up a little dialog asking for a small donation to this effect?

Something along the lines of:
> In order to make your use of OpenEmu safer, we want to make it compatible with Apple's Gatekeeper system.
> This will guarantee that your version of OpenEmu was properly packaged by the OpenEmu team and did not get tampered by malware.
>
> However, registering to the developer program requires a 99$ fee.
> So please consider donating: any amount you can contribute will make OpenEmu safer.
>
> Donations so far: xx $ out of 99$.

I would be very happy to donate with such an incentive. ;)

**Comment by robotoad** [comment]
Excluding files was obsoleted in version 2 code signing, and therefore all code and resources in a bundle be signed. From [OS X Code Signing In Depth](https://developer.apple.com/library/mac/technotes/tn2206/_index.html):

> Systems before OS X Mavericks v10.9 documented a signing feature (--resource-rules) to control which files in a bundle should be sealed by a code signature. This feature has been obsoleted for Mavericks. Code signatures made in Mavericks and later always seal all files in a bundle; there is no need to specify this explicitly any more. This also means that the Code Signing Resource Rules Path build setting in Xcode should no longer be used and should be left blank.

> It is thus no longer possible to exclude parts of a bundle from the signature. Bundles should be treated as read-only once they have been signed.

**Comment by nekoniaow** [comment]
Note that it is possible to have changing data files without invalidating the signature.

Moving the data which is modified from `info.plist` to another place where it makes more sense such as `Application Support\OpenEmu` is probably a good thing design wise.

From the link in the previous comment:
> * I write a receipt/license file/magic marker into my bundle after installation.
> This is no longer allowed.
> Write data into files outside the bundle as described earlier.
> Use the extended attribute mechanism described earlier to point your bundle at this data if necessary.
> Use an online service to store it and use the extended attribute mechanism to point at data there.

> * I store data in my bundle after I sign it.
> This is no longer allowed. You will need to locate this data elsewhere.
> If you write the data before your app runs for the first time, Gatekeeper will reject your app because the signature will be invalid.
> If you write that data after your app first runs, you're still breaking your signature. Gatekeeper will check your app again if your app is copied to another computer such that the new copy is quarantined.
> OS X APIs that rely on a valid identity will fail.
> In general, you should plan for future stricter runtime checks of code validity.

**Comment by clobber** [comment]
@nekoniaow Yeah...I think we get the picture, now please enough chatter regarding this.

I don't think you understand codesigning and why our [Info.plist is being modified](http://stackoverflow.com/questions/8542022/can-you-dynamically-assign-cfbundledocumenttypes-to-your-cocoa-application). You didn't know the terminology, costs associated and were wrong about bundle exclusions (it was even in the documentation you misinterpreted), so I think we'll let the developers on the project handle this.

At this point, users are trained to 'right click app > click open' when hit with the "unidentified developer" warning ever since OS X 10.7.5. This is not a big deal, as a lot of open source software is distributed unsigned, including but not limited to DeSmuME, ECWolf, EDuke32, Higan, RetroArch, OpenXcom, Mozilla's Servo alphas, etc.

>In order to make your use of OpenEmu safer, we want to make it compatible with Apple's Gatekeeper system.

Signing with an Apple Developer ID isn't going to keep you any safer from bugs in our app, us deleting all your Pokemon ROMs or protected against **[native code execution vulnerabilities](https://www.youtube.com/watch?v=Q3SOYneC7mU)** inherent in emulation.

As stated before your first reply in this issue, we currently have legitimate technical reasons for not signing which we may or may not work around. For now, the application is delivered via Sparkle which enforces SSL transfers and DSA signatures at the minimum.
